File: api/src/structs/get.rs
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize)]
pub struct SeriesNameList {
    pub name: String,
}

#[derive(Serialize, Deserialize)]
pub struct DebtSeriesPatchJson {
    pub debt_series: Vec<DebtSeries>,
    pub debt_pricing: Vec<DebtPricing>,
    pub debt_service: Vec<DebtService>,
}

#[derive(Serialize, Deserialize)]
pub struct DebtSeries {
    pub id: Option<i64>,
    pub series_name: String,
    pub is_tax_exempt: Option<bool>,
    pub par_amount: f64,
    pub premium: Option<f64>,
    pub cost_of_issuance: Option<f64>,
    pub created_at: Option<String>,
}

#[derive(Serialize, Deserialize)]
pub struct DebtPricing {
    pub id: Option<i64>, // optional for inserts
    pub series_id: i64,
    pub maturity_date: String, // YYYY-MM-DD
    pub amount: f64,
    pub coupon_rate: f64,
    pub yield_rate: f64,
    pub price: f64,
    pub premium_discount: Option<f64>,
    pub created_at: Option<String>,
}

#[derive(Serialize, Deserialize)]
pub struct DebtService {
    pub id: Option<i64>, // optional for inserts
    pub series_id: i64,
    pub payment_date: String, // YYYY-MM-DD
    pub principal: Option<f64>,
    pub interest: Option<f64>,
    pub created_at: Option<String>,
}

File: frontend/src/components/Constants/Get.ts
// Represents a simple series name
export interface SeriesNameList {
  name: string;
}

// Represents a patch request for debt series, pricing, and service
export interface DebtSeriesPatchJson {
  debt_series: DebtSeries[];
  debt_pricing: DebtPricing[];
  debt_service: DebtService[];
}

// Represents a debt series
export interface DebtSeries {
  id: number; // i64 → number
  series_name: string;
  is_tax_exempt?: boolean | null;
  par_amount: number;
  premium?: number | null;
  cost_of_issuance?: number | null;
  created_at?: string | null; // optional timestamp string
}

// Represents debt pricing information
export interface DebtPricing {
  id: number;
  series_id: number;
  maturity_date: string; // YYYY-MM-DD
  amount: number;
  coupon_rate: number;
  yield_rate: number;
  price: number;
  premium_discount?: number | null;
  created_at?: string | null;
}

// Represents debt service information
export interface DebtService {
  id: number;
  series_id: number;
  payment_date: string; // YYYY-MM-DD
  principal?: number | null;
  interest?: number | null;
  created_at?: string | null;
}
